### PR TITLE
Remove 0.75x downscaling of SVG icons

### PR DESCRIPTION
**Pull request by dvdmrtnz**
This PR removes the 0.75x downscaling of the SVG icons of POI in the current SVG→PNG workflow.  This will simplify work for icon designers as they will be able to use a 1px grid to design the icons that will match the final pixel grid of the icons in the app.

I have modified all the icons to take into account the removal of this resize, so they look the same size in the app. However, current icons will have to be tweaked to take advantage of this change and make them match the grid. That is, unless they were designed already with the previous 1pt=0.75px conversion in mind. In that case they will match the new grid just fine.

* Fixes section 3 of https://github.com/organicmaps/organicmaps/issues/4017

### CODE REVIEW

**Comment by dvdmrtnz** [comment]
> APK size has taken 200Ko

That's weird. How do you know that?

**Comment by Jean-BaptisteC** [comment]
Forget this :)
We have lost 20Ko (I have compare master fdroid beta and pr fdroid beta)

**Comment by pastk** [comment]
How do `viewBox="0 0 24 24" width="18" height="18"` or `viewBox="0 0 16 16" width="12" height="12"` work together? What's the difference between viewbox and width/height definitions?

**Comment by pastk** [comment]
I've noticed that dot icons `dot-m.svg` became smaller.
Now:
image

Before:
image

Didn't notice it with other icons.

Is there a way to ensure all icons are converted properly?

**Review by euf — APPROVED** [review]

**Comment by euf** [comment]
@dvdmrtnz Thank you so much for this amazing job! I checked a bunch of files manually, they seem to look fine both in Inkscape and in external SVG viewers. Could you please share how did you resize all these files? What software did you use? 

@pastk Thanks for noticing the `dot-m.svg` issue, I missed it 😔 

> How do `viewBox="0 0 24 24" width="18" height="18"` or `viewBox="0 0 16 16" width="12" height="12"` work together? What's the difference between viewbox and width/height definitions?

I'm no expert In SVG, but as far as I can tell, the following happens:
1. The `viewBox` describes a "canvas size" in some generic units, it goes `min-x min-y width height`. All further shape and curve coordinates are also in these generic units. These units do not have some physical dimension.
2. The `width` and `height` are specified in "user units", which are treated as `px` by default. They set up a "viewport", describing a physical size of the image. Then `viewBox` is rendered to fit inside these width and height.
3. Taking an SVG with `viewBox="0 0 24 24" width="24" height="24"` and changing it to `width="18" height="18"` while keeping original `viewBox` makes the whole image to be resized to 18x18px from 24x24px. All generic SVG units stay the same, so no need to edit files for recalculating shapes coordinates. Effectively the 0.75x scaling now happens at the moment of translating generic SVG units to user units before display. 

That's my impression after reading explainers [1], [2], [3] and playing around with tools [4], [5]. 
Feel free to correct my mistakes.

1. https://docs.aspose.com/svg/net/drawing-basics/coordinate-systems-and-units/#user-coordinate-system-and-svg-units
2. https://web.archive.org/web/20140119030353/https://software.intel.com/en-us/html5/blogs/viewbox-a-window-into-the-soul-of-svg
3. https://www.sarasoueidan.com/blog/svg-coordinate-systems/
5. https://www.sarasoueidan.com/demos/interactive-svg-coordinate-system/
6. https://www.svgviewer.dev

**Comment by dvdmrtnz** [comment]
> @dvdmrtnz Thank you so much for this amazing job! I checked a bunch of files manually, they seem to look fine both in Inkscape and in external SVG viewers. Could you please share how did you resize all these files? What software did you use?

It was quite a manual work. I used the "find and replace in all files" feature in sublime text. 

> @pastk Thanks for noticing the `dot-m.svg` issue, I missed it 😔
> 
> > How do `viewBox="0 0 24 24" width="18" height="18"` or `viewBox="0 0 16 16" width="12" height="12"` work together? What's the difference between viewbox and width/height definitions?
> 
> I'm no expert In SVG, but as far as I can tell, the following happens:
> 
>     1. The `viewBox` describes a "canvas size" in some generic units, it goes `min-x min-y width height`. All further shape and curve coordinates are also in these generic units. These units do not have some physical dimension.
> 
>     2. The `width` and `height` are specified in "user units", which are treated as `px` by default. They set up a "viewport", describing a physical size of the image. Then `viewBox` is rendered to fit inside these width and height.
> 
>     3. Taking an SVG with `viewBox="0 0 24 24" width="24" height="24"` and changing it to `width="18" height="18"` while keeping original `viewBox` makes the whole image to be resized to 18x18px from 24x24px. All generic SVG units stay the same, so no need to edit files for recalculating shapes coordinates. Effectively the 0.75x scaling now happens at the moment of translating generic SVG units to user units before display.
> 
> 
> That's my impression after reading explainers [1], [2], [3] and playing around with tools [4], [5]. Feel free to correct my mistakes.
> 
>     1. https://docs.aspose.com/svg/net/drawing-basics/coordinate-systems-and-units/#user-coordinate-system-and-svg-units
> 
>     2. https://web.archive.org/web/20140119030353/https://software.intel.com/en-us/html5/blogs/viewbox-a-window-into-the-soul-of-svg
> 
>     3. https://www.sarasoueidan.com/blog/svg-coordinate-systems/
> 
>     4. https://www.sarasoueidan.com/demos/interactive-svg-coordinate-system/
> 
>     5. https://www.svgviewer.dev

As fas as I know, everything you said here about svg dimensions is correct. The way I think about it is: 
* the viewbox uses the same coordinates as all the geometric elements inside the svg file, 
* the width/height sets the size of the image that will be declared to the outside world

**Comment by dvdmrtnz** [comment]
> I've noticed that dot icons `dot-m.svg` became smaller. Now: image
> 
> Before: image
> 
> Didn't notice it with other icons.

I have been taking a look at this and it seems that there is some problem with the width/height if they are not whole pixels but decimals. 

> Is there a way to ensure all icons are converted properly?

We should be able to compare the generated PNG skins to see if they match. However they icons seem to not be placed consistently. I will try to fix this. Also the decimal width/height are causing some discrepancies that make the icons not match exactly. I have to take a deeper look into this.

**Comment by pastk** [comment]
> We should be able to compare the generated PNG skins to see if they match. However the icons seem to not be placed consistently. I will try to fix this. Also the decimal width/height are causing some discrepancies that make the icons not match exactly. I have to take a deeper look into this.

Probably the order of icons in the skin is determined by their size and after that by? name maybe?. For sure it'll be good to make it consistent (if its not) and document it somewhere (a comment in the sources at least).

**Comment by euf** [comment]
> re decimal sizes - IMO should be alright to round them, no need for sub-pixel precision here
Agree. I believe it would be easier to manually fix possible issues with affected icons 
- `dot-m.svg`
- `subway-dalian-m.svg`
- `subway-sapporo-m.svg`
after this PR, while checking manually all other icons in terms of their pixel grid placement

**Comment by dvdmrtnz** [comment]
I have tweaked the icons size so they match the size of mdpi exactly. As you can see in the commits: I regenerated the icons before and after removing the 0.75x downscaling, and the mdpi skin does not appear in the second regeneration commit because it is exactly the same as before. 

@euf PTAL

**Comment by euf** [comment]
Amazing! I checked commits and compared png skins, everything looks just right 🔥

File: skin_generator/generator.cpp [hunk 1]
```diff
@@ -96,7 +96,7 @@ void SkinGenerator::ProcessSymbols(std::string const & svgDataDir,
           QSize defaultSize = m_svgRenderer.defaultSize();
 
           QSize symbolSize = symbolSizes[j];
-          QSize size = defaultSize * (symbolSize.width() / 24.0);
+          QSize size = defaultSize * (symbolSize.width() / 18.0);
```
Comment (vng) [inline skin_generator/generator.cpp:99]: Can we clarify here where the 18 constant comes from?
Is it something that m_svgRenderer.defaultSize is 18?
Should we still keep this transformation when we adjusted all the svg dimensions (in later commits)?

**Review by vng — APPROVED** [review]

File: skin_generator/generator.cpp (hunk 1 above)
Comment (pastk) [reply skin_generator/generator.cpp:99]: I don't understand it too and it'll be great to clarify this logic.

Not necessary in this PR.

(I wanted to go through the code and learn it myself, but don't have time now)

**Review by pastk — APPROVED** [review]
Did anyone test the last version in an app build?

File: skin_generator/generator.cpp (hunk 1 above)
Comment (dvdmrtnz) [reply skin_generator/generator.cpp:99]: As far as I know this is how this code works:
* `m_svgRenderer.defaultSize()` is the size of the SVG file (After the resize in this PR generally the -m files are 18x18 and the -s files are 12x12 ─ previously -m files were 24x24 and -s files were 16x16)
* `symbolSize` is the size of the PNG skins that we want to generate and they are defined in the script [generate_symbols.sh](https://github.com/organicmaps/organicmaps/blob/master/tools/unix/generate_symbols.sh#L82-L94): 18 (mdpi), 27 (hdpi), 36 (xhdpi)... 
* The `18.0` comes from the fact that the icons that should be represented as 18x18 in mdpi are also saved as 18x18 SVG files. Previously the constant `24.0` came from the fact that the icons that should be represented as 18x18 in mdpi were saved as 24x24 SVG files.

Comment (pastk) [reply skin_generator/generator.cpp:99]: Many thanks! Could you please convert this knowledge into the code comments?

Comment (dvdmrtnz) [reply skin_generator/generator.cpp]: Added comments

**Comment by patepelo** [comment]
Thanks for this. I couldn't understand it but it seems like results in a simplification for designing icons. I'll catch on it for later designs.

**Pull request merged**